### PR TITLE
Track key origins while merging multiple sources

### PR DESCRIPTION
**Pull request by akaIDIOT**

### CODE REVIEW

File: confidence/models.py [hunk 1]
```diff
@@ -34,7 +35,8 @@ class Configuration(Mapping):
     def __init__(self,
                  *sources: typing.Mapping[str, typing.Any],
                  separator: str = '.',
-                 missing: typing.Any = Missing.silent):
+                 missing: typing.Any = Missing.silent,
+                 origin: typing.Optional[str] = None):
```
Comment (akaIDIOT) [inline confidence/models.py:39]: `Origin`

**Comment by HHoofs** [comment]
Waarom?

**Comment by akaIDIOT** [comment]
> Waarom?
> Aka kan je me een voorbeeld geven hoe en wanneer dit handig is?

For debugging purposes mainly. When loading configuration content from multiple sources, a source with a higher precedence could override something defined in another file. Should some value not be what you had expected at runtime, it'd be useful to be able to see where that value is from. For example:

~~~~ yaml
# system defaults in /etc/example.yaml
example.service.port: 1234
~~~~

~~~~ yaml
# local configuration file next to the code, ./example.yaml
example.service.prt: 2345
~~~~

At runtime, one might expect `config.example.service.port` to result in port number 2345, but due to the typo, it'd be 1234. `confidence.why(config, 'example.service.port')` won't fix the typo, but would at least be able to tell you that that value was loaded from from `/etc/example.yaml`. Even without the typo, running the code that goes with this configuration from a different working directory might 'miss' `./example.yaml` completely, also potentially causing values to be different from what was expected by the user. 

In short: a way of being able to debug where something was loaded from, particularly for a more complex setup where many sources are combined into a single configuration namespace.

**Comment by HHoofs** [comment]
I see, I find it quite 'lenient' towards the user. Confidence is even for an user like me not that hard. Mistakes can only happen in a few places. So if this PR becomes too complex I would maybe hesitate (if it is for 'free' however, why not!)

For me, personally, it would be 'enough' to have a list with all the sources of an Configuration object (without the mapping). 

Most mistakes come from erroneous placement of the yamls. If I could quickly see which one I missed, it would be enough. For example when I use wsl on Windows as python interpreter I often forget that the user folder is no longer on my windows path...

**Comment by akaIDIOT** [comment]
After having played around with some logging / debug callbacks for the various loader methods, while not trivial, they are a lot simpler than the solution in this PR. As the proposition here is quite invasive and requires a lot of additional information to be tracked, choice requires a clearer picture of what something like this would actually be used for.

Inspired by `pip config debug`, ftr.

**Comment by akaIDIOT** [comment]
As this a) no longer fits the main branch and b) is terribly invasive for a debugging solution, let's look at the ideas from the last comment and see if we can have some component log the sources and overrides rather than always tracking everything.
